libvmaf/src/feature/cambi.c

```diff
@@ -810,25 +852,59 @@ static int cambi_score(VmafPicture *pics, uint32_t *mask_dp, uint16_t window_siz
     return 0;
 }
 
+static int preprocess_and_extract_cambi(CambiState *s, VmafPicture *pic, double *score, bool is_src) {
+    int width = is_src ? s->src_width : s->enc_width;
+    int height = is_src ? s->src_height : s->enc_height;
+    int window_size = is_src ? s->src_window_size : s->window_size;
+    int num_diffs = 1 << s->max_log_contrast;
+    
+    int err = cambi_preprocessing(pic, &s->pics[0], width, height);
+    if (err) return err;
+
+    err = cambi_score(s->pics, s->mask_dp, window_size, s->topk, num_diffs, s->tvi_for_diff, s->c_values, 
+                      s->c_values_histograms, s->filter_mode_histogram, s->filter_mode_buffer, score, width, height);
+    if (err) return err;
+
+    return 0;
+}
+
+static double combine_dist_src_scores(double dist_score, double src_score) {
+    return MAX(0, dist_score - src_score);
+}
+
 static int extract(VmafFeatureExtractor *fex,
                    VmafPicture *ref_pic, VmafPicture *ref_pic_90,
                    VmafPicture *dist_pic, VmafPicture *dist_pic_90,
                    unsigned index, VmafFeatureCollector *feature_collector) {
-    (void)ref_pic;
     (void)ref_pic_90;
     (void)dist_pic_90;
 
     CambiState *s = fex->priv;
-
-    int err = cambi_preprocessing(dist_pic, &s->pics[0]);
+    double dist_score;
+    int err = preprocess_and_extract_cambi(s, dist_pic, &dist_score, false);
     if (err) return err;
 
-    double score;
-    const uint16_t num_diffs = 1<<s->max_log_contrast;
-    err = cambi_score(s->pics, s->mask_dp, s->window_size, s->topk, num_diffs, s->tvi_for_diff, s->c_values, s->c_values_histograms, s->filter_mode_histogram, s->filter_mode_buffer, &score);
-    if (err) return err;
+    if (s->full_reference) {
+        double src_score;
+        int err = preprocess_and_extract_cambi(s, ref_pic, &src_score, true);
+        if (err) return err;
+
+        err = vmaf_feature_collector_append(feature_collector, "cambi_source", src_score, index);
+        if (err) return err;
+
+        double combined_score = combine_dist_src_scores(dist_score, src_score);
+        err = vmaf_feature_collector_append(feature_collector, "cambi_full_reference", combined_score, index);
+        if (err) return err;
+    }
+    else {
```

Comment:
If the `full_reference` flag is not set, this block returns -1 on the `cambi_source` and `cambi_full_reference` metrics. This is to avoid conditionally returning features. There may be a cleaner way to do this, please tell me if that is the case.